File: NDL_Category/NDL_Category/TestGestureViewController.m
//
//  TestGestureViewController.m
//  NDL_Category
//
//  Created by dzcx on 2018/12/4.
//  Copyright © 2018 ndl. All rights reserved.
//

#import "TestGestureViewController.h"

@interface TestGestureViewController () <UIScrollViewDelegate, UIGestureRecognizerDelegate>

@property (nonatomic, weak) UIView *bgView;

@property (nonatomic, strong) UIScrollView *scrollView;

@property (nonatomic, strong) UIImageView *imageView;

@property (nonatomic, strong) UIPanGestureRecognizer *pan;

@end

@implementation TestGestureViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIView *bgView = [[UIView alloc] initWithFrame:CGRectMake(20, 80, 200, 180)];
    bgView.backgroundColor = [UIColor redColor];
    [self.view addSubview:bgView];
    self.bgView = bgView;
    
    
    self.scrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(20, 80, 200, 180)];
    self.scrollView.bounces = YES;
    self.scrollView.delegate = self;
    // ###contentSize > scrollView.size才会触发###
    [self.scrollView.panGestureRecognizer addTarget:self action:@selector(scrollViewPan:)];
    self.scrollView.backgroundColor = [UIColor redColor];
    [self.view addSubview:self.scrollView];
    
    self.imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 280, 260)];
//    self.imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 180, 160)];
    self.imageView.userInteractionEnabled = YES;// 必须写
    self.imageView.backgroundColor = [UIColor cyanColor];
    [self.scrollView addSubview:self.imageView];
    
    self.pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(imageViewPan:)];
    [self.pan addTarget:self action:@selector(imageViewSecondPan:)];// 可以添加多个target/action
    self.pan.delegate = self;
    [self.imageView addGestureRecognizer:self.pan];
    
    // contentSize > scrollView.size
    // 下面都不写 imageViewPan默认先触发
    // scrollViewPan优先触发
//    [self.pan requireGestureRecognizerToFail:self.scrollView.panGestureRecognizer];
    // imageViewPan优先触发 (panGestureRecognizer不触发，也就不滚动了)
//    [self.scrollView.panGestureRecognizer requireGestureRecognizerToFail:self.pan];
    
    // contentSize < scrollView.size
    // 下面都不写 imageViewPan默认先触发
    // imageViewPan优先触发 因为scrollViewPan没有触发
//    [self.pan requireGestureRecognizerToFail:self.scrollView.panGestureRecognizer];
//    imageViewPan优先触发
//    [self.scrollView.panGestureRecognizer requireGestureRecognizerToFail:self.pan];
    
    self.scrollView.contentSize = self.imageView.size;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    // frame改变了
    NSLog(@"frame = %@ tran = %@", NSStringFromCGRect(self.scrollView.frame), NSStringFromCGAffineTransform(self.scrollView.transform));// frame = {{120, 170}, {100, 90}}
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    // 同时写 只用后面的
//    self.scrollView.transform = CGAffineTransformMakeScale(0.5, 0.5);// size = {100, 90} halfSize = {50, 45}
//    self.scrollView.transform = CGAffineTransformMakeTranslation(50, 45);
    // 同时有效果
    self.scrollView.transform = CGAffineTransformMake(0.5, 0, 0, 0.5, 50, 45);
    
}

- (void)scrollViewPan:(UIPanGestureRecognizer *)gesture
{
    NSLog(@"scrollViewPan");
}

// contentSize > scrollView.size才会触发
//- (void)scrollViewDidScroll:(UIScrollView *)scrollView
//{
//    NSLog(@"TestGestureViewController scrollViewDidScroll");
//}

//
- (void)imageViewPan:(UIPanGestureRecognizer *)gesture
{
    NSLog(@"imageViewPan");
}

- (void)imageViewSecondPan:(UIPanGestureRecognizer *)gesture
{
    NSLog(@"imageViewSecondPan");
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    // 只有contentSize > scrollView.size才会触发 YES表示两个都执行（scrollViewPan && imageViewPan）
    if (gestureRecognizer == self.pan && otherGestureRecognizer == self.scrollView.panGestureRecognizer) {
        NSLog(@"gestureRecognizer = %@ otherGestureRecognizer = %@", gestureRecognizer, otherGestureRecognizer);
        
        return YES;// NO的话只执行gestureRecognizer所对应的 即imageViewPan
    }
    
    return NO;
}

@end

File: SwiftDemo/SwiftDemo/TestGesture/TestGestureViewController.swift
//
//  TestGestureViewController.swift
//  SwiftDemo
//
//  Created by ndl on 2020/2/20.
//  Copyright © 2020 dzcx. All rights reserved.
//

import UIKit

class TestGestureViewController: UIViewController, UIGestureRecognizerDelegate, UIScrollViewDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.lightGray

        let wrapperView = UIView(frame: CGRect(x: 0, y: 100, width: self.view.frame.width, height: self.view.frame.height - 200))
        wrapperView.backgroundColor = UIColor.white
        self.view.addSubview(wrapperView)
        
        let pan = UIPanGestureRecognizer()
        pan.delegate = self
        pan.addTarget(self, action: #selector(onPanned(_:)))
        
        wrapperView.addGestureRecognizer(pan)
        
        let tap = UITapGestureRecognizer()
        tap.delegate = self
        tap.addTarget(self, action: #selector(onTapped(_:)))
        wrapperView.addGestureRecognizer(tap)
        
        // test button
//        let button = UIButton(type: .custom)
//        button.backgroundColor = UIColor.red
//        button.frame = CGRect(x: 10, y: 10, width: 50, height: 50)
//        button.addTarget(self, action: #selector(buttonDidClicked), for: .touchUpInside)
//        wrapperView.addSubview(button)
        
        // test scrollView
        let scrollView = UIScrollView()
        scrollView.delegate = self
        scrollView.bounces = false
        scrollView.frame = wrapperView.bounds
        scrollView.contentSize = CGSize(width: self.view.frame.width * 2.0, height: wrapperView.frame.height)
        scrollView.backgroundColor = UIColor.yellow
        wrapperView.addSubview(scrollView)
        
//
//        print("scrollView.panGestureRecognizer.delegate = \(scrollView.panGestureRecognizer.delegate)")
    }
    
    @objc func onPanned(_ gesture: UIGestureRecognizer) {
        print("onPanned: \(gesture.self)")
    }
    
    @objc func onTapped(_ gesture: UIGestureRecognizer) {
        print("onTapped: \(gesture.self)")
    }
    
    @objc func buttonDidClicked() {
        print("buttonDidClicked")
    }
    
    // MARK: 总结
    /**
     1.
     对于 UIButton，UISwitch，UISegmentedControl，UIStepper、UIPageControl 进行单击操作，如果父视图有轻敲手势需要识别，依然会按照响应链来处理，先响应这些控件的单击事件，这仅适用于与控件的默认操作重叠的手势识别

     
     */
    
    // MARK: UIGestureRecognizerDelegate
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        print("gestureRecognizer = \(gestureRecognizer) otherGestureRecognizer = \(otherGestureRecognizer)")
        
        return true
    }
    
//    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
//        print("touch.view = \(touch.view)")
//        if let flag = touch.view?.isKind(of: UIScrollView.self), flag {
//            print("===scrollview")
//            return false
//        }
//
//        return true
//    }
    

}
